packages/teams-js/src/internal/validOrigins.ts

```diff
@@ -0,0 +1,125 @@
+import { validOriginsCdnEndpoint, validOriginsFallback } from './constants';
+import { GlobalVars } from './globalVars';
+import { getLogger } from './telemetry';
+import { inServerSideRenderingEnvironment, isValidHttpsURL } from './utils';
+
+let validOriginsCache: string[] = [];
+const validateOriginLogger = getLogger('validateOrigin');
+
+export async function prefetchOriginsFromCDN(): Promise<void> {
+  await getValidOriginsListFromCDN();
+}
+
+function isValidOriginsCacheEmpty(): boolean {
+  return validOriginsCache.length !== 0;
+}
+
+async function getValidOriginsListFromCDN(): Promise<string[]> {
+  if (isValidOriginsCacheEmpty()) {
+    return validOriginsCache;
+  }
+  if (!inServerSideRenderingEnvironment()) {
+    return fetch(validOriginsCdnEndpoint)
+      .then((response) => {
+        if (!response.ok) {
+          throw new Error('Invalid Response from Fetch Call');
+        }
+        return response.json().then((validOriginsCDN) => {
+          if (validateValidOriginsFromCDN(JSON.stringify(validOriginsCDN))) {
+            validOriginsCache = validOriginsCDN.validOrigins;
+            return validOriginsCache;
+          } else {
+            throw new Error('Valid Origins List Is Invalid');
+          }
+        });
+      })
+      .catch((e) => {
+        validateOriginLogger('validOrigins fetch call to CDN failed with error: %s. Defaulting to fallback list', e);
+        validOriginsCache = validOriginsFallback;
+        return validOriginsCache;
+      });
+  } else {
+    validOriginsCache = validOriginsFallback;
+    return validOriginsFallback;
+  }
+}
+
+function validateValidOriginsFromCDN(validOriginsJSON: string): boolean {
```

Comment:
>boolean [](http://example.com/codeflow?start=64&length=7)

It looks like you want to indicate errors by returning `false` rather than throwing.

As it is written, `JSON.parse` will throw if the string isn't valid JSON (and that's not in your narrowly-scoped try/catch block). Consider expanding the scope of your try catch to be the entire function in this case -- or add a second try/catch around JSON.parse.